File: drivers/dart/test/helpers_test.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:reddb/reddb.dart';
import 'package:test/test.dart';

class _FakeCall {
  _FakeCall(this.sql, this.params);
  final String sql;
  final List<Object?>? params;
}

class _FakeQuerier implements Querier {
  _FakeQuerier({List<Object?>? replies, List<Object?>? errors})
      : _replies = List.from(replies ?? const []),
        _errors = List.from(errors ?? const []);

  final List<_FakeCall> calls = [];
  final List<Object?> _replies;
  final List<Object?> _errors;

  @override
  Future<Uint8List> query(String sql, {List<Object?>? params}) async {
    calls.add(_FakeCall(sql, params));
    final err = _errors.isEmpty ? null : _errors.removeAt(0);
    if (err != null) throw err;
    final reply = _replies.isEmpty ? null : _replies.removeAt(0);
    return Uint8List.fromList(utf8.encode(jsonEncode(reply ?? {})));
  }
}

Uint8List _reply(Object? v) =>
    Uint8List.fromList(utf8.encode(jsonEncode(v)));

void main() {
  group('kvPath', () {
    test('quotes namespaced keys', () {
      expect(kvPath('kv_default', 'corpus:version'),
          "kv_default.'corpus:version'");
    });
    test('preserves dots and slashes inside the key', () {
      expect(kvPath('kv_default', 'a/b.c'), "kv_default.'a/b.c'");
    });
    test('rejects bad collection', () {
      expect(
        () => kvPath('bad-name!', 'k'),
        throwsA(isA<InvalidArgument>()),
      );
    });
  });

  group('KV', () {
    test('set emits exact key path and value literal', () async {
      final q = _FakeQuerier(replies: [<String, Object?>{}]);
      await Helpers(q).kv.set('characters:hansel', 'ok');
      final sql = q.calls.first.sql;
      expect(sql, contains("kv_default.'characters:hansel'"));
      expect(sql, contains("= 'ok'"));
    });

    test('escapes single quotes in value literal', () async {
      final q = _FakeQuerier(replies: [<String, Object?>{}]);
      await Helpers(q).kv.set('k', "o'reilly");
      expect(q.calls.first.sql, contains("= 'o''reilly'"));
    });

    test('get returns value or null on miss', () async {
      final q = _FakeQuerier(replies: [
        {
          'rows': [
            {'value': 'v'},
          ],
        },
        {'rows': <Object?>[]},
      ]);
      final kv = Helpers(q).kv;
      expect(await kv.get('k'), 'v');
      expect(await kv.get('k2'), isNull);
    });

    test('exists is true when get returns a value', () async {
      final q = _FakeQuerier(replies: [
        {
          'rows': [
            {'value': 'v'},
          ],
        },
        {'rows': <Object?>[]},
      ]);
      final kv = Helpers(q).kv;
      expect((await kv.exists('k')).exists, isTrue);
      expect((await kv.exists('k2')).exists, isFalse);
    });

    test('list filters by prefix without rewriting the SQL', () async {
      final q = _FakeQuerier(replies: [
        {
          'rows': [
            {'key': 'a:1', 'value': 1},
            {'key': 'b:1', 'value': 2},
            {'key': 'a:2', 'value': 3},
          ],
        },
      ]);
      final res = await Helpers(q).kv.list(prefix: 'a:');
      expect(res.items, hasLength(2));
      expect(res.items.map((r) => r['key']), ['a:1', 'a:2']);
      expect(q.calls.first.sql, isNot(contains('a:')));
    });

    test('list rejects negative limit', () async {
      expect(
        () => Helpers(_FakeQuerier()).kv.list(limit: -1),
        throwsA(isA<InvalidArgument>()),
      );
    });
  });

  group('Queue', () {
    test('push emits priority and JSON payload', () async {
      final q = _FakeQuerier(replies: [
        {'affected': 1},
      ]);
      await Helpers(q).queue.push('jobs', {'id': 1}, priority: 5);
      final sql = q.calls.first.sql;
      expect(sql, startsWith('QUEUE PUSH jobs '));
      expect(sql, contains('PRIORITY 5'));
      expect(sql, contains('{"id":1}'));
    });

    test('len returns int from row', () async {
      final q = _FakeQuerier(replies: [
        {
          'rows': [
            {'len': 3},
          ],
        },
      ]);
      expect(await Helpers(q).queue.len('jobs'), 3);
    });

    test('pop returns payload list', () async {
      final q = _FakeQuerier(replies: [
        {
          'rows': [
            {'payload': 'a'},
            {'payload': 'b'},
          ],
        },
      ]);
      final out = await Helpers(q).queue.pop('jobs', count: 2);
      expect(out, ['a', 'b']);
    });

    test('pop rejects negative count', () async {
      expect(
        () => Helpers(_FakeQuerier()).queue.pop('jobs', count: -1),
        throwsA(isA<InvalidArgument>()),
      );
    });

    test('push rejects invalid identifier', () async {
      expect(
        () => Helpers(_FakeQuerier()).queue.push('bad-name!', 'x'),
        throwsA(isA<InvalidArgument>()),
      );
    });
  });

  group('Documents', () {
    test('insert returns rid envelope', () async {
      final q = _FakeQuerier(replies: [
        {'rows': <Object?>[], 'affected': 0},
        {
          'rows': [
            {
              'rid': 'doc-1',
              'body': {'name': 'alice'},
            },
          ],
          'affected': 1,
        },
      ]);
      final out =
          await Helpers(q).documents.insert('people', {'name': 'alice'});
      expect(out.affected, 1);
      expect(out.rid, 'doc-1');
      expect(out.item!['rid'], 'doc-1');
    });

    test('get raises NotFound on missing row', () async {
      final q = _FakeQuerier(replies: [
        {'rows': <Object?>[]},
      ]);
      expect(
        () => Helpers(q).documents.get('people', 'doc-1'),
        throwsA(isA<NotFound>()),
      );
    });

    test('patch rejects JSON-pointer-style paths', () async {
      expect(
        () => Helpers(_FakeQuerier())
            .documents
            .patch('people', 'doc-1', {'a/b': 1}),
        throwsA(isA<InvalidArgument>()),
      );
    });

    test('list orders by rid ASC by default', () async {
      final q = _FakeQuerier(replies: [
        {
          'rows': [
            {'rid': 'a'},
            {'rid': 'b'},
          ],
        },
      ]);
      final out = await Helpers(q).documents.list('people');
      expect(out.items, hasLength(2));
      expect(q.calls.first.sql, contains('ORDER BY rid ASC'));
    });

    test('passes through "collection already exists" error', () async {
      final q = _FakeQuerier(
        replies: [
          <String, Object?>{},
          {
            'rows': [
              {'rid': 'x'},
            ],
            'affected': 1,
          },
        ],
        errors: [Exception('collection already exists'), null],
      );
      final out = await Helpers(q).documents.insert('people', {'a': 1});
      expect(out.rid, 'x');
    });
  });

  group('response parsing', () {
    test('nested result.affected envelope is honoured by delete', () async {
      final q = _FakeQuerier(replies: [
        {
          'result': {'affected': 7},
        },
      ]);
      final out = await Helpers(q).queue.purge('jobs');
      expect(out.affected, 7);
      expect(out.deleted, isTrue);
    });
  });

  // --- SDK Helper Spec v1.0 ---------------------------------------------

  group('spec v1.0', () {
    test('exposes HELPER_SPEC_VERSION = "1.0"', () {
      expect(Helpers.helperSpecVersion, '1.0');
    });

    test('documents.patch rejects empty patch with INVALID_ARGUMENT', () async {
      final q = _FakeQuerier();
      expect(
        () => Helpers(q).documents.patch('people', 'doc-1', {}),
        throwsA(isA<InvalidArgument>()),
      );
      expect(q.calls, isEmpty,
          reason: 'empty patch must reject before issuing any query');
    });

    test('documents.delete of missing rid returns deleted=false (not error)',
        () async {
      final q = _FakeQuerier(replies: [
        {'affected': 0},
      ]);
      final out = await Helpers(q).documents.delete('people', 'no-such-rid');
      expect(out.affected, 0);
      expect(out.deleted, isFalse);
    });

    test('documents.delete of present rid returns deleted=true', () async {
      final q = _FakeQuerier(replies: [
        {'affected': 1},
      ]);
      final out = await Helpers(q).documents.delete('people', 'doc-1');
      expect(out.affected, 1);
      expect(out.deleted, isTrue);
    });

    test('kv.delete of missing key returns deleted=false', () async {
      final q = _FakeQuerier(replies: [
        {'affected': 0},
      ]);
      final out = await Helpers(q).kv.delete('nope');
      expect(out.affected, 0);
      expect(out.deleted, isFalse);
    });

    test('documents.bulkInsert empty is no-op', () async {
      final q = _FakeQuerier();
      final out = await Helpers(q).documents.bulkInsert('people', const []);
      expect(out.affected, 0);
      expect(out.rids, isEmpty);
      expect(q.calls, isEmpty);
    });

    test('documents.bulkInsert preserves per-row rid order', () async {
      final q = _FakeQuerier(replies: [
        // _ensureCollection
        <String, Object?>{},
        // first insert
        {
          'rows': [
            {'rid': 'r-1'},
          ],
          'affected': 1,
        },
        // second insert
        {
          'rows': [
            {'rid': 'r-2'},
          ],
          'affected': 1,
        },
      ]);
      final out = await Helpers(q).documents.bulkInsert('people', [
        {'i': 1},
        {'i': 2},
      ]);
      expect(out.affected, 2);
      expect(out.rids, ['r-1', 'r-2']);
    });

    test('queues alias and queues.create emit CREATE QUEUE IF NOT EXISTS',
        () async {
      final q = _FakeQuerier(replies: [<String, Object?>{}]);
      await Helpers(q).queues.create('jobs');
      expect(q.calls.first.sql, 'CREATE QUEUE IF NOT EXISTS jobs');
    });

    test('queues.create rejects bad identifier', () async {
      expect(
        () => Helpers(_FakeQuerier()).queues.create('bad-name!'),
        throwsA(isA<InvalidArgument>()),
      );
    });

    test('tx.begin / commit / rollback emit BEGIN / COMMIT / ROLLBACK',
        () async {
      final q = _FakeQuerier(
          replies: [<String, Object?>{}, <String, Object?>{}, <String, Object?>{}]);
      final tx = Helpers(q).tx();
      await tx.begin();
      await tx.commit();
      await tx.begin();
      await tx.rollback();
      expect(q.calls.map((c) => c.sql), ['BEGIN', 'COMMIT', 'BEGIN', 'ROLLBACK']);
    });

    test('tx.run commits on success', () async {
      final q = _FakeQuerier(replies: [
        <String, Object?>{}, // BEGIN
        <String, Object?>{}, // body insert
        <String, Object?>{}, // COMMIT
      ]);
      final tx = Helpers(q).tx();
      await tx.run((t) async {
        await q.query('SOMETHING');
      });
      expect(q.calls.map((c) => c.sql), ['BEGIN', 'SOMETHING', 'COMMIT']);
    });

    test('tx.run rolls back and re-throws on error', () async {
      final q = _FakeQuerier(
        replies: [<String, Object?>{}, <String, Object?>{}, <String, Object?>{}],
        errors: [null, Exception('boom'), null],
      );
      final tx = Helpers(q).tx();
      await expectLater(
        tx.run((t) async => q.query('SOMETHING')),
        throwsA(isA<Exception>()),
      );
      expect(q.calls.map((c) => c.sql), ['BEGIN', 'SOMETHING', 'ROLLBACK']);
    });

    test('tx.run rejects nested run with INVALID_ARGUMENT', () async {
      final q = _FakeQuerier(replies: [<String, Object?>{}]);
      final tx = Helpers(q).tx();
      await expectLater(
        tx.run((outer) async {
          await outer.run((_) async {});
        }),
        throwsA(isA<InvalidArgument>()),
      );
    });
  });
}

File: drivers/js/test/helpers.test.mjs
/**
 * SDK Helper Spec — helper-surface unit tests (no binary required).
 *
 * These exercise the spec-conformant additions made for issue #559 against a
 * fake RPC client so they run everywhere `node --test` does:
 *   - HELPER_SPEC_VERSION / db.helperSpecVersion (spec §14)
 *   - kv.set alias + kv.delete `{ affected, deleted }` envelope (spec §5)
 *   - queues.create idempotent SQL + `db.queues` alias (spec §6)
 *   - tx.begin / commit / rollback + tx.run commit / rollback / nested reject (spec §7)
 */

import { test } from 'node:test'
import assert from 'node:assert/strict'

import {
  RedDB,
  RedDBError,
  KvClient,
  QueueClient,
  HELPER_SPEC_VERSION,
} from '../src/index.js'

function fakeDb(handler) {
  const calls = []
  const client = {
    async call(method, params) {
      calls.push({ method, params })
      return handler?.(method, params) ?? { rows: [] }
    },
    async close() {},
  }
  return { db: new RedDB(client, { transport: 'embedded' }), calls }
}

function fakeClient(responder) {
  const calls = []
  const client = {
    async call(method, params) {
      calls.push({ method, params })
      return responder?.(method, params) ?? { rows: [] }
    },
  }
  return { client, calls }
}

test('HELPER_SPEC_VERSION is 1.0 and exposed on the db', () => {
  assert.equal(HELPER_SPEC_VERSION, '1.0')
  const { db } = fakeDb()
  assert.equal(db.helperSpecVersion, '1.0')
})

test('kv.set is an alias for put and emits the same SQL', async () => {
  const { client, calls } = fakeClient()
  const kv = new KvClient(client, 'settings')
  await kv.set('characters:hansel', 'crumbs')
  assert.equal(calls[0].method, 'query')
  assert.equal(calls[0].params.sql, "KV PUT settings.'characters:hansel' = 'crumbs'")
})

test('kv.delete returns the { affected, deleted } envelope', async () => {
  const { client } = fakeClient((_, params) => (
    params.sql.endsWith('.gone') ? { affected: 0 } : { affected: 1 }
  ))
  const kv = new KvClient(client, 'settings')
  assert.deepEqual(await kv.delete('present'), { affected: 1, deleted: true })
  assert.deepEqual(await kv.delete('gone'), { affected: 0, deleted: false })
})

test('queues.create emits idempotent CREATE QUEUE IF NOT EXISTS', async () => {
  const { client, calls } = fakeClient()
  const queue = new QueueClient(client)
  await queue.create('jobs')
  assert.equal(calls[0].params.sql, 'CREATE QUEUE IF NOT EXISTS jobs')
})

test('queues.create rejects bad identifiers before issuing a query', () => {
  const { client, calls } = fakeClient()
  const queue = new QueueClient(client)
  assert.throws(() => queue.create('bad-name'), (err) => {
    assert.ok(err instanceof RedDBError)
    assert.equal(err.code, 'INVALID_QUEUE_NAME')
    return true
  })
  assert.equal(calls.length, 0)
})

test('db.queues aliases db.queue', () => {
  const { db } = fakeDb()
  assert.ok(db.queues instanceof QueueClient)
  assert.equal(db.queues, db.queue)
})

test('tx.begin / commit issue BEGIN then COMMIT and clear tx state', async () => {
  const { db, calls } = fakeDb()
  const tx = db.tx()
  await tx.begin()
  assert.equal(db.inTransaction, true)
  await tx.commit()
  assert.equal(db.inTransaction, false)
  assert.deepEqual(calls.map((c) => c.params.sql), ['BEGIN', 'COMMIT'])
})

test('tx.rollback issues ROLLBACK and clears tx state', async () => {
  const { db, calls } = fakeDb()
  const tx = db.tx()
  await tx.begin()
  await tx.rollback()
  assert.equal(db.inTransaction, false)
  assert.deepEqual(calls.map((c) => c.params.sql), ['BEGIN', 'ROLLBACK'])
})

test('tx.commit / rollback without an open transaction reject with INVALID_ARGUMENT', async () => {
  const { db } = fakeDb()
  await assert.rejects(
    () => db.tx().commit(),
    (err) => err instanceof RedDBError && err.code === 'INVALID_ARGUMENT',
  )
  await assert.rejects(
    () => db.tx().rollback(),
    (err) => err instanceof RedDBError && err.code === 'INVALID_ARGUMENT',
  )
})

test('tx.run commits on success and returns the callback value', async () => {
  const { db, calls } = fakeDb((method) => (
    method === 'insert' ? { affected: 1, id: 7 } : { rows: [] }
  ))
  const value = await db.tx().run(async (tx) => {
    await tx.insert('users', { name: 'Ada' })
    return { ok: true }
  })
  assert.deepEqual(value, { ok: true })
  assert.deepEqual(calls.map((c) => c.params?.sql ?? c.method), ['BEGIN', 'insert', 'COMMIT'])
  assert.equal(db.inTransaction, false)
})

test('tx.run rolls back and re-throws when the callback throws', async () => {
  const { db, calls } = fakeDb((method) => (
    method === 'insert' ? { affected: 1, id: 8 } : { rows: [] }
  ))
  await assert.rejects(
    () => db.tx().run(async (tx) => {
      await tx.insert('users', { name: 'Grace' })
      throw new Error('boom')
    }),
    /boom/,
  )
  assert.deepEqual(calls.map((c) => c.params?.sql ?? c.method), ['BEGIN', 'insert', 'ROLLBACK'])
  assert.equal(db.inTransaction, false)
})

test('nested tx.run rejects with INVALID_ARGUMENT and the outer tx still commits', async () => {
  const { db, calls } = fakeDb()
  await db.tx().run(async () => {
    await assert.rejects(
      () => db.tx().run(async () => 'nested'),
      (err) => err instanceof RedDBError && err.code === 'INVALID_ARGUMENT',
    )
  })
  assert.deepEqual(calls.map((c) => c.params.sql), ['BEGIN', 'COMMIT'])
})
